File: sfn/SfnWorkflow.cs
using sfn_ut.sfn.lambda;
using sfn_ut.sfn.wait;
using sfn_ut.sfn.nested;
using sfn_ut.sfn.await;
using sfn_ut.sfn.error;
using sfn_ut.sfn.parallel;

namespace sfn_ut.sfn;

public enum Mode {
  Test,
  ExtractMetadata
}

public abstract class SfnWorkflow<PAYLOAD> {
  public List<SfnStep> Steps { get; set; } = [];
  public SfnErrorHandler<PAYLOAD> ErrorHandler { get; set; }
  public string WorkflowName { get; private set; }

  public abstract PAYLOAD Run(PAYLOAD input);

  public SfnWorkflow(SfnErrorHandler<PAYLOAD> errorHandler) {
    ErrorHandler = errorHandler;
    WorkflowName = GetType().FullName!;
  }

  internal void HandleErrorInternal(PAYLOAD payload, Exception e) {
    // Do logging, assertions, etc
    ErrorHandler.Execute(new SfnErrorHandlerInput<PAYLOAD>(payload, e.Message));
  }

  public PAYLOAD Lambda(SfnLambda<PAYLOAD> lambda, PAYLOAD input) {
    Steps.Add(lambda);

    try {
      return lambda.Execute(input);
    }
    catch (Exception e) {
      HandleErrorInternal(input, e);
      throw;
    }
  }

  public void Wait(string stepName, TimeSpan duration) {
    Steps.Add(new SfnWait<PAYLOAD>(stepName, duration));
  }

  public PAYLOAD LambdaLoop(
    SfnLambda<PAYLOAD> lambda,
    SfnIsComplete<PAYLOAD> isComplete,
    PAYLOAD payload
    ) {
    SfnLambdaWithLoop<PAYLOAD> lambdaWithAwait = new(lambda, isComplete);
    Steps.Add(lambdaWithAwait);

    try {
      return lambdaWithAwait.Execute(payload);
    }
    catch (Exception e) {
      HandleErrorInternal(payload, e);
      throw;
    }
  }

  public PAYLOAD Parallel(List<SfnStepWithPayload<PAYLOAD>> steps, SfnParallelResultsAssembler<PAYLOAD> assembler, PAYLOAD payload) {
    SfnParallel<PAYLOAD> parallel = new(steps, assembler);

    Steps.Add(parallel);
    Steps.Add(assembler);

    try {
      return parallel.Execute(payload);
    }
    catch (Exception e) {
      HandleErrorInternal(payload, e);
      throw;
    }
  }

  public PAYLOAD NestedSfn<CHILD>(
    SfnTransform<PAYLOAD, CHILD> transform,
    SfnWorkflow<CHILD> nested,
    SfnMerge<PAYLOAD, CHILD> merge,
    PAYLOAD payload) {

    SfnWorkflowInvocation<PAYLOAD, CHILD> nestedInvocation = new(transform, nested, merge);

    Steps.Add(transform);
    Steps.Add(nestedInvocation);
    Steps.Add(merge);

    try {
      return nestedInvocation.Execute(payload);
    }
    catch (Exception e) {
      HandleErrorInternal(payload, e);
      throw;
    }
  }
}

File: java-sfn-unit-test/src/main/java/com/sfnut/sfn/SfnWorkflow.java
package com.sfnut.sfn;

import com.sfnut.sfn.error.SfnErrorHandlerInput;
import com.sfnut.sfn.lambda.SfnLambda;
import com.sfnut.sfn.lambdawithloop.SfnIsComplete;
import com.sfnut.sfn.lambdawithloop.SfnLambdaWithLoop;
import com.sfnut.sfn.nested.SfnMerge;
import com.sfnut.sfn.nested.SfnMergeInput;
import com.sfnut.sfn.nested.SfnTransform;
import com.sfnut.sfn.nested.SfnWorkflowInvocation;
import com.sfnut.sfn.parallel.SfnParallel;
import com.sfnut.sfn.parallel.SfnParallelResultsAssembler;
import com.sfnut.sfn.wait.SfnWait;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;

public abstract class SfnWorkflow<PAYLOAD> {
    @JsonProperty("Steps")
    private List<SfnStep> steps = new ArrayList<>();
    
    @JsonProperty("ErrorHandler")
    private SfnErrorHandler<PAYLOAD> errorHandler;
    
    @JsonProperty("WorkflowName")
    private String workflowName;

    public SfnWorkflow(SfnErrorHandler<PAYLOAD> errorHandler) {
        this.errorHandler = errorHandler;
        this.workflowName = this.getClass().getName();
    }

    public abstract PAYLOAD run(PAYLOAD input);

    protected void handleErrorInternal(PAYLOAD payload, Exception e) {
        // Do logging, assertions, etc
        errorHandler.execute(new SfnErrorHandlerInput<>(payload, e.getMessage()));
    }

    public PAYLOAD lambda(SfnLambda<PAYLOAD> lambda, PAYLOAD input) {
        steps.add(lambda);

        try {
            return lambda.execute(input);
        } catch (Exception e) {
            handleErrorInternal(input, e);
            throw e;
        }
    }

    public void waitStep(String stepName, Duration duration) {
        steps.add(new SfnWait<>(stepName, duration));
    }

    public PAYLOAD lambdaLoop(SfnLambda<PAYLOAD> lambda,
                             SfnIsComplete<PAYLOAD> isComplete,
                             PAYLOAD payload) {
        SfnLambdaWithLoop<PAYLOAD> lambdaWithLoop = new SfnLambdaWithLoop<>(lambda, isComplete);
        steps.add(lambdaWithLoop);

        try {
            return lambdaWithLoop.execute(payload);
        } catch (Exception e) {
            handleErrorInternal(payload, e);
            throw e;
        }
    }

    public PAYLOAD parallel(List<SfnStepWithPayload<PAYLOAD>> stepList,
                           SfnParallelResultsAssembler<PAYLOAD> assembler,
                           PAYLOAD payload) {
        SfnParallel<PAYLOAD> parallel = new SfnParallel<>(stepList, assembler);

        steps.add(parallel);
        steps.add(assembler);

        try {
            return parallel.execute(payload);
        } catch (Exception e) {
            handleErrorInternal(payload, e);
            throw e;
        }
    }

    public <CHILD> PAYLOAD nestedSfn(SfnTransform<PAYLOAD, CHILD> transform,
                                    SfnWorkflow<CHILD> nested,
                                    SfnMerge<PAYLOAD, CHILD> merge,
                                    PAYLOAD payload) {
        SfnWorkflowInvocation<PAYLOAD, CHILD> nestedInvocation = 
            new SfnWorkflowInvocation<>(transform, nested, merge);

        steps.add(transform);
        steps.add(nestedInvocation);
        steps.add(merge);

        try {
            return nestedInvocation.execute(payload);
        } catch (Exception e) {
            handleErrorInternal(payload, e);
            throw e;
        }
    }

    // Getters and setters
    public List<SfnStep> getSteps() {
        return steps;
    }

    public void setSteps(List<SfnStep> steps) {
        this.steps = steps;
    }

    public SfnErrorHandler<PAYLOAD> getErrorHandler() {
        return errorHandler;
    }

    public void setErrorHandler(SfnErrorHandler<PAYLOAD> errorHandler) {
        this.errorHandler = errorHandler;
    }

    public String getWorkflowName() {
        return workflowName;
    }

    public void setWorkflowName(String workflowName) {
        this.workflowName = workflowName;
    }
}
